xapian-applications/omega/postprocess.py

```diff
@@ -0,0 +1,118 @@
+# Postprocess click data files for obtaining the final clickstream log file
+# and generate Query file for Xapian Letor module from it.
+#
+# Copyright (C) 2017 Vivek Pal
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+# USA
+
+from ast import literal_eval
+from sys import argv
+
+import csv
+
+script_name, search_log, clicks_log = argv
+
+try:
+    with open(search_log, 'r') as s:
+        with open('log/clickstream.log', 'w+') as f:
+            reader = csv.reader(s, delimiter = '\t')
+            writer = csv.writer(f, delimiter = '\t', quoting = csv.QUOTE_MINIMAL)
+
+            for row in reader:
+                # Skip rows with empty Query string
+                if row[1] == '':
+                    continue
+                # Remove an extra ',' character at the end of the Hitlist string
+                hits = row[2]
+                hits = hits[:-2]
+                row[2] = hits
+
+                # Add Clicks column
+                row.append(row[2])
+
+                # Convert Hitlist from string to list
+                hits = row[2]
+                hits = hits.strip().split(',')
+                row[2] = hits
+
+                # Convert Clicks from string to list
+                clicks = row[4]
+                clicks = clicks.strip().split(',')
+                row[4] = clicks
+
+                # Initialise Clicks list elements to zero
+                for i in range(len(row[4])):
+                    row[4][i] = str(0)
+
+                writer.writerow(row)
```

Comment:
> However, I'm not sure why you bother writing this to an interim file, rather than collapsing this preprocessing with the subsequent loop.

At first, I was trying to achieve the same i.e. eliminating the need of having an interim file (`clickstream.log`) but it didn't seem possible to me because there was no way to read and write into `final.log` at the same time.

This is what I tried: opened `search.log` in read mode and `final.log` in write mode to add "Clicks" column to it with all values set to `0` of length same as hitlist and closed it. And, to update those values, we need to be able to read as well as write (row with updated clicks values) which isn't quite possible.

Also, I think it is safer (from data loss perspective) to use independent files to generate `final.log` rather than trying to edit `final.log` itself. What do you think?